test/spmd/tensor/test_megatron_example.py

```diff
@@ -96,18 +115,57 @@ def aggregate_output(outputs):
     return dist_mod
 
 
+def shard_self_attn(m, device_type):
+    pg = dist.distributed_c10d._get_default_group()
+    start_idx = 0
+    device_mesh = DeviceMesh(
+        device_type,
+        list(range(start_idx, start_idx + pg.size())),
+        dim_groups=[pg],
+    )
+    col_wise_sharding = [Shard(0)]
+    row_wise_sharding = [Shard(1)]
+    replicate = [Replicate()]
+    m.qkv.weight = torch.nn.Parameter(
+        distribute_tensor(m.qkv.weight, device_mesh, col_wise_sharding)
+    )
+    m.proj.weight = torch.nn.Parameter(
+        distribute_tensor(m.proj.weight, device_mesh, row_wise_sharding)
+    )
+    if m.qkv.bias is not None:
+        m.qkv.bias = torch.nn.Parameter(
+            distribute_tensor(m.qkv.bias, device_mesh, col_wise_sharding)
+        )
+    if m.proj.bias is not None:
+        m.proj.bias = torch.nn.Parameter(
+            distribute_tensor(m.proj.bias, device_mesh, replicate)
+        )
+    m = _replicate_input_tensor(m, device_mesh, replicate)
+    m.proj = _aggregate_local_tensor(m.proj)
+    m.qkv.weight.register_hook(functools.partial(_gradient_hook, m.qkv.weight))
+    if m.qkv.bias is not None:
+        m.qkv.bias.register_hook(functools.partial(_gradient_hook, m.qkv.bias))
+    m.proj.weight.register_hook(
+        functools.partial(_gradient_hook, m.proj.weight)
+    )
+    if m.proj.bias is not None:
+        m.proj.bias.register_hook(
+            functools.partial(_gradient_hook, m.proj.bias)
+        )
+
+
 class DistTensorMegatronTest(DistTensorTestBase):
     @with_comms
-    def test_simple_megatron_e2e(self):
+    def test_mlp_megatron_e2e(self):
         LR = 0.5
         inp_size = [5, 10]
         torch.manual_seed(0)
         inp = torch.rand(*inp_size, device=self.device_type)
         torch.manual_seed(5)
-        model = SimpleModel(self.device_type)
+        model = MLPModule(self.device_type)
         torch.manual_seed(5)
-        model_tp = SimpleModel(self.device_type)
-        model_tp = shard_module(model_tp, self.device_type)
+        model_tp = MLPModule(self.device_type)
+        shard_mlp(model_tp, self.device_type)
```

Comment:
Initialize the optimizer here before running the forward pass?

Reply:
Out of curiosity, is there any reason for that?

Reply:
I only mention it as a way to improve readability. Wrote down what I meant in more detail below.